### PR TITLE
bin/brew: move logic to Library.

### PR DESCRIPTION
**Pull request by mikemcquaid**
`bin/brew` is getting a bit bloated and it's annoying to open all of `HOMEBREW_PREFIX` in your editor of choice and have it search through the `Cellar`. Let's instead keep the bare minimum in `bin/brew` and
move the rest to `Library`.

### CODE REVIEW

**Comment by mikemcquaid** [comment]
To make review easier: no code has been changed here, just moved around (and `chdir` duplicated).

**Comment by UniqMartin** [comment]
> `bin/brew` is getting a bit bloated and it's annoying to open all of `HOMEBREW_PREFIX` in your editor of choice and have it search through the `Cellar`.

Not a criticism of this PR, but wouldn't using distinct paths for `HOMEBREW_PREFIX` and `HOMEBREW_REPOSITORY` (something Homebrew already supports natively) be a solution for this use case?

File: bin/brew [hunk 1]
```diff
@@ -41,149 +18,9 @@ HOMEBREW_PREFIX="$(chdir "$(dirname "$(dirname "$HOMEBREW_BREW_FILE")")" && pwd
...
-  # at the same time as being run.
-  #
-  # Hide shellcheck complaint:
-  # shellcheck source=/dev/null
-  source "$HOMEBREW_BASH_COMMAND"
-  { "homebrew-$HOMEBREW_COMMAND" "$@"; exit $?; }
-else
-  exec "$HOMEBREW_RUBY_PATH" -W0 "$HOMEBREW_LIBRARY/brew.rb" "$HOMEBREW_COMMAND" "$@"
-fi
+exec /bin/bash "$HOMEBREW_LIBRARY/brew.sh" "$@"
```
Comment (UniqMartin) [inline bin/brew]: I think using `source "$HOMEBREW_LIBRARY/brew.sh"` should give the same effect with less overhead.

File: bin/brew [hunk 2]
```diff
@@ -41,149 +18,9 @@ HOMEBREW_PREFIX="$(chdir "$(dirname "$(dirname "$HOMEBREW_BREW_FILE")")" && pwd
...
-    HOMEBREW_RUBY_PATH="$(which ruby)"
-  fi
-fi
-
 export HOMEBREW_BREW_FILE
-export HOMEBREW_RUBY_PATH
 export HOMEBREW_PREFIX
 export HOMEBREW_REPOSITORY
 export HOMEBREW_LIBRARY
-export HOMEBREW_CELLAR
```
Comment (UniqMartin) [inline bin/brew]: If we source `Library/brew.sh` instead of executing it in a new shell, these `export`s can be moved there, too.

**Comment by UniqMartin** [comment]
Two comments, but otherwise I'm happy with this change.

We might want to make `brew.sh` more modular in a follow-up PR by moving some of its parts into separate files (no idea yet where those should live). Even if we stick with a single file, we might want to transform it to a structure as sketched out below:

```shell
function force-utf8() {
  # Force UTF-8 to avoid encoding issues for users with broken locale settings.
  if [[ "$(locale charmap 2> /dev/null)" != "UTF-8" ]]
  then
    export LC_ALL="en_US.UTF-8"
  fi
}

setup-cellar() {
  # Where we store built products; /usr/local/Cellar if it exists,
  # otherwise a Cellar relative to the Repository.
  if [[ -d "$HOMEBREW_PREFIX/Cellar" ]]
  then
    HOMEBREW_CELLAR="$HOMEBREW_PREFIX/Cellar"
  else
    HOMEBREW_CELLAR="$HOMEBREW_REPOSITORY/Cellar"
  fi
}

handle-path-commands() {
  case "$*" in
    --prefix) echo "$HOMEBREW_PREFIX"; exit 0 ;;
    --cellar) echo "$HOMEBREW_CELLAR"; exit 0 ;;
    --repository|--repo) echo "$HOMEBREW_REPOSITORY"; exit 0 ;;
  esac
}

# (more functions with similar structure …)

force-utf8
setup-cellar
handle-path-commands "$@" # handles plain --cellar/--prefix/--repository and exits quickly
check-prefix
clean-environment
setup-ruby
# (more checks and preparatory work …)
handle-command "$@" # invokes either Bash or Ruby command and exits
```

IMO, this will make the code more readable and the chunk at the bottom communicates what the script does much more succinctly.

**Comment by mikemcquaid** [comment]
I'd be :+1: on a structure like that @UniqMartin. Any objections to the change as-is or is it :ok: from your perspective to :ship:?

**Comment by UniqMartin** [comment]
> Any objections to the change as-is or is it :ok: from your perspective to :ship:?

Would be lovely if you could address the two code comments I left (or comment on them if you disagree) as I'm strongly in favor of using `source` instead of `exec`.

**Comment by mikemcquaid** [comment]
@UniqMartin Yeh, I will, just wanted to check for sure you're :ok: otherwise.

**Comment by UniqMartin** [comment]
@mikemcquaid I'm generally in favor of the proposed split. Just would prefer for it to be *exec*uted (sorry for the bad pun) subtly differently …

**Comment by mistydemeo** [comment]
> bin/brew is getting a bit bloated and it's annoying to open all of HOMEBREW_PREFIX in your editor of choice and have it search through the Cellar.

I'm happy to move stuff out of `bin/brew`, but re: this part - maybe worth checking if your editor has a way to open projects rather than entire directories at the CLI? `brew edit` supports this for TextMate and Sublime Text, opening `bin/brew`, the README, and a few specific directories without opening the entirety of `HOMEBREW_PREFIX`, without opening the Cellar.

**Comment by mikemcquaid** [comment]
@mistydemeo Yeh, my editor does; that was a somewhat simplified example as when I say "my editor" I also mean "and grep/ack/find/other unixy tools that are a pain to add specific white/black lists too". Good points though.

**Comment by mikemcquaid** [comment]
Addressed @UniqMartin's comments and pull, thanks folks.

**Pull request closed without merging**